### DISCUSSION TITLE
handshaking报错问题

### DISCUSSION THREAD

**Discussion by LoserXj** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#60] [Q&A]
![报错1](https://user-images.githubusercontent.com/103046776/210861877-0d023dd4-3265-4238-9d93-863d3c7127ad.png)
请问对于这个报错是可能是什么原因呢，有时候debug又能过，在这里不会报错。因为这个是握手定理，我的peer使用命令行测试（使用随机实现丢包）都是可以实现的，而且peer使用可靠性测试那里的命令都完全可以实现，不知道为啥这个握手定理这里会报错。之前测过握手和可靠数据拥塞都能过，但是我改了peer后面握手这里就一直过不了，但是peer应该不会影响这里呀，所以只能求救了

**Comment by TANSixu** [comment]
Hi,请问可以提供一下运行这个test失败后的checkerlog以及peer1 log， peer2 log的部分嘛（就前面几十行就行）

**Reply by LoserXj** [reply]
![报错2](https://user-images.githubusercontent.com/103046776/210936508-74089262-fcc6-445a-adfc-270c7c701ba6.png)
![报错3](https://user-images.githubusercontent.com/103046776/210936521-3db01b40-9c12-42fb-a4aa-2a03b8b90545.png)
![报错4](https://user-images.githubusercontent.com/103046776/210936536-6ea57e57-ff97-4503-8331-f0720322a1ed.png)
![报错5](https://user-images.githubusercontent.com/103046776/210936545-d7285e46-11b3-4602-83e2-dee72209fecf.png)
![报错6](https://user-images.githubusercontent.com/103046776/210936553-aa2afe67-c9dd-4061-84c0-cc2a975741c3.png)
之前回复错了，不好意思

**Reply by TANSixu** [reply]
> 之前回复错了，不好意思

没事没事 不过刚刚我也回答到下面去了，记得往下翻一下噢～ 😂

**Comment by LoserXj** [comment]
![报错2](https://user-images.githubusercontent.com/103046776/210932709-9e112927-aa0d-40a7-9a02-228ae04640de.png)
![报错3](https://user-images.githubusercontent.com/103046776/210932721-cb6fc417-54a7-47de-a432-42a488c5f579.png)
![报错4](https://user-images.githubusercontent.com/103046776/210932732-c38dd6c5-9b45-4b29-9e20-3bc7c7e20630.png)
![报错5](https://user-images.githubusercontent.com/103046776/210932745-f1e4c127-e088-4f60-a9e4-bc07188bc488.png)
![报错6](https://user-images.githubusercontent.com/103046776/210932752-25b23cc8-5fee-45e7-a922-4ec6fcfebe97.png)

**Reply by TANSixu** [reply]
是由于peer启动速度慢了，在peer1发出whohas是peer2还没有完全启动，如log的时间戳所示。同学可以检查一下grader.py中 class peerproc的start peer方法：
```
    def start_peer(self):
        if self.timeout:
            cmd = f"python3 -u {self.peer_file_loc} -p {self.node_map_loc} -c {self.haschunk_loc} -m {self.max_transmit} -i {self.id} -t {self.timeout}"
        else:
            cmd = f"python3 -u {self.peer_file_loc} -p {self.node_map_loc} -c {self.haschunk_loc} -m {self.max_transmit} -i {self.id}"
        
        peer_stdout_file = open(f"log/peer{self.id}.stdout", "w")
        self.process = subprocess.Popen(cmd.split(" "), stdin=subprocess.PIPE,stdout=peer_stdout_file, stderr=peer_stdout_file,text=True, bufsize=1, universal_newlines=True)
        # ensure peer is running
        time.sleep(1) 
```
下面的time sleep就是等待peer启动的时间。我们此前收到另外两位同学的反馈，0.1秒的等待时间在import了其他包后不足以让peer完成启动，于是在 #48 中将0.1s统一更新为了1s。请查看此处是否更新。如果1s仍然不够完成启动，可以自行更改为更长时间。
